src/ParallelAlgorithms/Amr/Criteria/Persson.cpp

```diff
@@ -0,0 +1,116 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "ParallelAlgorithms/Amr/Criteria/Persson.hpp"
+
+#include <array>
+#include <cstddef>
+
+#include "DataStructures/ApplyMatrices.hpp"
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Matrix.hpp"
+#include "Domain/Amr/Flag.hpp"
+#include "NumericalAlgorithms/Spectral/Filtering.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/Numeric.hpp"
+#include "Utilities/SetNumberOfGridPoints.hpp"
+
+namespace amr::Criteria {
+
+template <size_t Dim>
+double persson_smoothness_indicator(
+    const gsl::not_null<DataVector*> filtered_component_buffer,
+    const DataVector& tensor_component, const Mesh<Dim>& mesh,
+    const size_t dimension, const size_t num_highest_modes) {
+  // Zero out the lowest modes in the given dimension
+  static const Matrix identity{};
+  auto matrices = make_array<Dim>(std::cref(identity));
+  gsl::at(matrices, dimension) = Spectral::filtering::zero_lowest_modes(
+      mesh.slice_through(dimension),
+      mesh.extents(dimension) - num_highest_modes);
+  set_number_of_grid_points(filtered_component_buffer,
+                            mesh.number_of_grid_points());
+  apply_matrices(filtered_component_buffer, matrices, tensor_component,
+                 mesh.extents());
+  // Take the L2 norm over all grid points
+  return sqrt(alg::accumulate(square(*filtered_component_buffer), 0.) /
+              filtered_component_buffer->size());
+}
+
+template <size_t Dim>
+std::array<double, Dim> persson_smoothness_indicator(
+    const DataVector& tensor_component, const Mesh<Dim>& mesh,
+    const size_t num_highest_modes) {
+  std::array<double, Dim> result{};
+  DataVector buffer{};
+  for (size_t d = 0; d < Dim; ++d) {
+    gsl::at(result, d) = persson_smoothness_indicator(
+        make_not_null(&buffer), tensor_component, mesh, d, num_highest_modes);
+  }
+  return result;
+}
+
+namespace Persson_detail {
+
+template <size_t Dim>
+void max_over_components(const gsl::not_null<std::array<Flag, Dim>*> result,
+                         const gsl::not_null<DataVector*> buffer,
+                         const DataVector& tensor_component,
+                         const Mesh<Dim>& mesh, const size_t num_highest_modes,
+                         const double relative_tolerance,
+                         const double absolute_tolerance,
+                         const double coarsening_factor) {
+  for (size_t d = 0; d < Dim; ++d) {
+    // Skip this dimension if we have already decided to refine it
+    if (gsl::at(*result, d) == Flag::IncreaseResolution) {
```

Comment:
Is `Split` not higher than `IncreaseResolution`?

Reply:
Ohh good catch, this should be `Split`.